### PR TITLE
Nucleus detection job

### LINKED ISSUE #153: Set up a nucleus detection job type
Use HistomicsTK (perhaps running in a container) to run a job to produce a nucleus ID pixel map image. This is an image with an integer value at each pixel encoding a numeric ID for the nucleus found at that location (`0` indicates "no pixel", while positive integers indicate uniquely identified nuclei).

In case my description of the file is wrong, whatever HistomicsTK actually does is to be taken as that description instead.

@manthey will be available to help get this issue done.

### PR DESCRIPTION
**Pull request by annehaley**
Resolves #153 

This job saves an array of dictionaries to the metadata of the output dataset. Each dictionary in the array represents one detected nucleus and comes with the following properties: 
```
{'Identifier.CentroidX': 372.4708520179372,
  'Identifier.CentroidY': 753.0,
  'Identifier.WeightedCentroidX': 371.538301233572,
  'Identifier.WeightedCentroidY': 751.3681220799301,
  'Identifier.Xmax': 393.0,
  'Identifier.Xmin': 356.0,
  'Identifier.Ymax': 769.0,
  'Identifier.Ymin': 735.0,
  'Label': 253.0,
  'Nucleus.Gradient.Canny.Mean': 0.2914798206278027,
  'Nucleus.Gradient.Canny.Sum': 130.0,
  'Nucleus.Gradient.Mag.HistEnergy': 0.18244887289106962,
  'Nucleus.Gradient.Mag.HistEntropy': 1.9060868159979303,
  'Nucleus.Gradient.Mag.Kurtosis': 1.1076810827786652,
  'Nucleus.Gradient.Mag.Mean': 33.96154913366462,
  'Nucleus.Gradient.Mag.Skewness': 1.2492234614579725,
  'Nucleus.Gradient.Mag.Std': 28.660294970497844,
  'Nucleus.Haralick.ASM.Mean': 0.006990539685474394,
  'Nucleus.Haralick.ASM.Range': 0.001949213750726088,
  'Nucleus.Haralick.Contrast.Mean': 21.367818592586396,
  'Nucleus.Haralick.Contrast.Range': 15.770061082754587,
  'Nucleus.Haralick.Correlation.Mean': 0.8811204237370258,
  'Nucleus.Haralick.Correlation.Range': 0.08690155212565231,
  'Nucleus.Haralick.DifferenceEntropy.Mean': 3.172526826395739,
  'Nucleus.Haralick.DifferenceEntropy.Range': 0.47473664492813716,
  'Nucleus.Haralick.DifferenceVariance.Mean': 0.0038460005118468523,
  'Nucleus.Haralick.DifferenceVariance.Range': 0.0013714254805384306,
  'Nucleus.Haralick.Entropy.Mean': 8.090542090749222,
  'Nucleus.Haralick.Entropy.Range': 0.32981335884097795,
  'Nucleus.Haralick.IDM.Mean': 0.393372923485391,
  'Nucleus.Haralick.IDM.Range': 0.06929654412392888,
  'Nucleus.Haralick.IMC1.Mean': -0.29313223585251297,
  'Nucleus.Haralick.IMC1.Range': 0.06825293263992988,
  'Nucleus.Haralick.IMC2.Mean': 0.967867424039129,
  'Nucleus.Haralick.IMC2.Range': 0.021249085453542538,
  'Nucleus.Haralick.SumAverage.Mean': 37.730752357842135,
  'Nucleus.Haralick.SumAverage.Range': 0.436815807404038,
  'Nucleus.Haralick.SumEntropy.Mean': 5.692367186319881,
  'Nucleus.Haralick.SumEntropy.Range': 0.04588099269859214,
  'Nucleus.Haralick.SumOfSquares.Mean': 89.80843550467458,
  'Nucleus.Haralick.SumOfSquares.Range': 0.7034500124560736,
  'Nucleus.Haralick.SumVariance.Mean': 337.8659234261121,
  'Nucleus.Haralick.SumVariance.Range': 12.996394007015624,
  'Nucleus.Intensity.HistEnergy': 0.13501176375957688,
  'Nucleus.Intensity.HistEntropy': 2.158366399214893,
  'Nucleus.Intensity.IQR': 109.0,
  'Nucleus.Intensity.Kurtosis': -1.5268149112964684,
  'Nucleus.Intensity.MAD': 52.5,
  'Nucleus.Intensity.Max': 154.0,
  'Nucleus.Intensity.Mean': 66.70627802690584,
  'Nucleus.Intensity.MeanMedianDiff': 8.206278026905835,
  'Nucleus.Intensity.Median': 58.5,
  'Nucleus.Intensity.Min': 0.0,
  'Nucleus.Intensity.Skewness': 0.15572138630667204,
  'Nucleus.Intensity.Std': 53.68652832708666,
  'Orientation.Orientation': 0.900410953745495,
  'Shape.Circularity': 0.2451683204887551,
  'Shape.Eccentricity': 0.8111664232186716,
  'Shape.EquivalentDiameter': 23.829914749152643,
  'Shape.Extent': 0.35453100158982515,
  'Shape.FSD1': 0.018545,
  'Shape.FSD2': 0.0019304,
  'Shape.FSD3': 0.0037173,
  'Shape.FSD4': 0.0075223,
  'Shape.FSD5': 0.3461223,
  'Shape.FSD6': 0.1325198,
  'Shape.FractalDimension': 1.2573466861883336,
  'Shape.HuMoments1': 0.26069859174844245,
  'Shape.HuMoments2': 0.016338280722474483,
  'Shape.HuMoments3': 0.00268592177001432,
  'Shape.HuMoments4': 0.0012468557573488363,
  'Shape.HuMoments5': 1.052845577933778e-06,
  'Shape.HuMoments6': 0.00010372010081738058,
  'Shape.HuMoments7': 2.0243442918483396e-06,
  'Shape.MinorMajorAxisRatio': 0.5848153844100092,
  'Shape.Solidity': 0.5334928229665071,
  'Shape.WeightedHuMoments1': 0.0050793409582471204,
  'Shape.WeightedHuMoments2': 3.34082266669627e-06,
  'Shape.WeightedHuMoments3': 2.270975026039596e-08,
  'Shape.WeightedHuMoments4': 2.547854175530759e-09,
  'Shape.WeightedHuMoments5': -4.159598574399326e-18,
  'Shape.WeightedHuMoments6': 5.008520972959511e-13,
  'Shape.WeightedHuMoments7': 1.8928993686342807e-17,
  'Size.Area': 446.0,
  'Size.ConvexHullArea': 836.0,
  'Size.MajorAxisLength': 37.232229556849845,
  'Size.MinorAxisLength': 21.77398064073085,
  'Size.Perimeter': 151.19595949289334}
```

### CODE REVIEW

**Comment by waxlamp** [comment]
The job is outputting results such as this:

`{"bbox": [0, 3445, 17, 3465], "centroid": [6.908088235294118, 3454.591911764706]}`

This format is great for visualizing the location of detected nuclei on the original image, but it doesn't seem well suited for submitting each detected nucleus for subsequent computations such as shape statistics, entropy, etc. Am I wrong about this, or is there some extra thing we could be doing here, like computing a "nucleus ID image" (where each pixel is either 0 for "no nucleus" or some specific integer value, indicating the presence of a specific nucleus at that location)?

This question is largely for @manthey and @annehaley, but glad to hear from others as well.

**Comment by manthey** [comment]
> like computing a "nucleus ID image" (where each pixel is either 0 for "no nucleus" or some specific integer value, indicating the presence of a specific nucleus at that location)?

It sounds like you want to generate a pixelmap with the nucleus id.  See the HistomicsTK pixelmap task for how that can be done.

I haven't looked at what you are doing at all, but in the HistomicsTK nuclei detection job, we have a bunch of scikit-image region properties that are discarded, keeping only the bounding box or polygon.  If you want those, you should have the blob finder save those results.

**Comment by waxlamp** [comment]
@annehaley, could you look into what it would take to include those skimage region properties?

**Comment by annehaley** [comment]
I found `skimage.measure.find_contours` which will return a list of x,y points forming a polygon around the mask sent into it. [4f6c430](https://github.com/atlascope/atlascope/pull/160/commits/4f6c430ec70c5d304396b5a15fb461e27339389d) uses this to get polygons for each identified nucleus, but this takes a while. I changed the job in the populate script to send in the smaller image, `HTA9_1_BA_F_ROI04`, which takes about 5 minutes to do the detection. For the larger image, `HTA9_1_BA_F_ROI02`, it took >10 minutes. 
This may be acceptable for now, but I'm wondering if there's a skimage function we can use to do this for all labels in a multi-label mask, rather than sending in a mask to the function for each detected nucleus.

**Comment by annehaley** [comment]
Here are some example visualizations of the generated polygons:

image
image

File: atlascope/core/job_types/nucleus_detection.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        im_nuclei_seg_mask, min_nucleus_area
+    ).astype(int)
+
+    # a list of polygons outlining the identified nuclei
+    # each polygon is a list of n points of the form [r, c]
+    # polygons = [
+    # switch coordinate order to get (x, y) format
+    # [[point[1], point[0]] for point in poly_points]
+    # for poly_points in skimage.measure.find_contours(im_nuclei_seg_mask, 0.5)
+    # ]
```
Comment (waxlamp) [inline atlascope/core/job_types/nucleus_detection.py]: Let's get rid of this commented block if we don't need this functionality (I don't think we do).

File: atlascope/core/job_types/nucleus_detection.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        job.complete = True
+        job.save()
+    except Exception as e:
+        print('FAILURE!')
+        print(e)
+        job.failure = str(e)
+        job.save()
+
+    print()
+    print()
```
Comment (waxlamp) [inline atlascope/core/job_types/nucleus_detection.py]: Are these needed?

**Review by waxlamp — APPROVED** [review]
This looks good to me (though, for the record, we discussed on 2022-05-25 an immediate update to this branch that does a more Djangoic job of modeling this analytic output data). I'll approve this so that you can merge as soon as any necessary cleanup (see inline comments) is done.

File: atlascope/core/job_types/nucleus_detection.py (hunk 1 above)
Comment (annehaley) [reply atlascope/core/job_types/nucleus_detection.py]: Removed in [28cddda](https://github.com/atlascope/atlascope/pull/160/commits/28cddda859ac9b14cbeae604e330bc361a580243)

File: atlascope/core/job_types/nucleus_detection.py (hunk 2 above)
Comment (annehaley) [reply atlascope/core/job_types/nucleus_detection.py]: Removed in [28cddda](https://github.com/atlascope/atlascope/pull/160/commits/28cddda859ac9b14cbeae604e330bc361a580243)

**Pull request merged**